### ISSUE TITLE
Stack overflow on large subject (~80k characters)

### ISSUE THREAD

**Issue by leni536** [hanickadot/compile-time-regular-expressions#30]
I have the following target:
```c++
auto re = "a+b"_ctre
```

The subject is 80000 characters of the character `'a'` passed in standard input then stored in an std::string. I ran it on Linux with 8192 KiB stack size, compiled with both g++ and clang++ with -O3. I get a stack overflow.

The full source of the test program:
```c++
#include <ctre.hpp>
#include <iostream>
#include <sstream>


int main() {
	using namespace ctre::literals;

	constexpr auto re_ctre = "a+b"_ctre;

	std::stringstream buf;
	buf << std::cin.rdbuf();

	const auto str = buf.str();

	volatile auto ctre_match = bool(re_ctre.match(str));

	return 0;
}
```

For certain patterns it's impossible to match in constant space, for these patterns it's guaranteed to have a stack overflow with large enough input.
I think it should be possible to match against the above specific pattern using constant space though.
I wonder how a functional style metaprogramming heavy library can avoid this. I think range-v3 can easily have the same problem.

**Comment by hanickadot** [comment]
Don't use greedy cycle, use lazy one or possessive:
greedy: `a+b`
lazy: `a+?b`
possessive: `a++b`

only greedy needs recursion for backtracking, other cycles doesn't grow stack.

**Comment by junyer** [comment]
> I think it should be possible to match against the above specific pattern using constant space though.

The BitState execution engine in RE2 recently became capable of that via run length encoding of stack entries.

**Comment by hanickadot** [comment]
From now on CTRE detects if first(greedy cycle content) is mutually exlusive with first(rest of regex).  If so, it changes the greedy cycle into possessive.  Which means no recursion.

This optimization is disable for backtracking and unicode properties (which are not currently enabled in master)

https://compiler-explorer.com/z/ndj589 

Fixed in 42e84d253903525cc5f6cf3bc65a1bbbb027c56c and also released new version 2.6.

**Comment by junyer** [comment]
Although RE2 does not support possessive quantifiers, it does have a somewhat similar optimisation that avoids pushing alternatives that cannot match the current input byte. (There is a little more to it than that, mind you, and it works for the NFA execution engine as well, not just the BitState execution engine.)
